### ISSUE TITLE
Make TaylorSeries work with DoubleDouble

### ISSUE THREAD

**Issue by dpsanders** [JuliaDiff/TaylorSeries.jl#56]
There are a string of conversion errors currently.

**Comment by lbenet** [comment]
Could you be more specific on this? I haven't worked with DoubleDluble...

**Comment by dpsanders** [comment]
```
julia> using TaylorSeries, DoubleDouble

julia> t = taylor1_variable(Double{Float64}, 10)
 Double(1.0, 0.0)
 - value: 1 t + 𝒪(t¹¹)

julia> t + t^2
 Double(1.0, 0.0)
 - value: 1 t + Double(1.0, 0.0)
 - value: 1 t² + 𝒪(t¹¹)

julia> (t+1)*(t+1)
 Double(1.0, 0.0)
 - value: 1 + Double(2.0, 0.0)
 - value: 2 t + Double(1.0, 0.0)
 - value: 1 t² + 𝒪(t¹¹)

julia> sin(t)
ERROR: MethodError: Cannot `convert` an object of type DoubleDouble.Double{Float64} to an object of type AbstractFloat
This may have arisen from a call to the constructor AbstractFloat(...),
since type constructors fall back to convert methods.
 in sincos(::TaylorSeries.Taylor1{DoubleDouble.Double{Float64}}) at /Users/dpsanders/.julia/v0.5/TaylorSeries/src/Taylor1.jl:718
 in sin(::TaylorSeries.Taylor1{DoubleDouble.Double{Float64}}) at /Users/dpsanders/.julia/v0.5/TaylorSeries/src/Taylor1.jl:703
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

This is after I fixed one missing conversion in DoubleDouble; it's possible that the problem is there, but I don't understand where the convert to AbstractFloat is happening in the TaylorSeries code.

**Comment by lbenet** [comment]
The problem is related to the fact that `sin(Double(0.0))`, as well as other elementary functions, are not defined for `Double`s; see also https://github.com/simonbyrne/DoubleDouble.jl/issues/8. This is needed by `TaylorSeries`, since the very first coefficient of a Taylor series is the function (`sin`) evaluated around the point you expand, in this case, `sin(Double(0.0))`.

**Comment by dpsanders** [comment]
Sorry, I didn't make clear that this was after I defined a convert for DoubleDouble that allows `zero(Double)` to work:
```
convert{T}(::Type{Double{T}}, x::Real) = convert(Double{T}, float(x))

```

**Comment by lbenet** [comment]
Thanks, good for the record.

**Comment by dpsanders** [comment]
The problem is actually that `sin` etc. are not yet defined on `Double`s. Things like `(1+t)^2` already just work.

**Comment by dpsanders** [comment]
However, when calculating `f(t)` where `f` is  `exp` or `sin` expanded around `0`, it *should not* be necessary to calculate `f(0)`?

**Comment by lbenet** [comment]
I disagree. You require to know $f(x_0)$ if you are developing around $x_0$. The case of `sin(x)` around zero is trivial, as well as for `cos`, `tan`, `exp`, etc. However, you may want to expand something like `sin(f(x))` around x=0, such that `f(0)` is not zero, which imposes the necessity of knowing `sin(f(0))`. Since we do not know it a priori, we cannot proceed.

In general (as it is mentioned above) the recurrence relations work but assume that $f(x_0)$ is somehow computable.

**Comment by dpsanders** [comment]
My point was that `sin t` should look at the 0th-order coefficient t[0].
If `t[0]==0`, then it should use the known result without calling `sin(t[0])` explicitly;
if it is non-zero, then it has to call `sin(t[0])`.

**Comment by lbenet** [comment]
I understood this. Yet, you could argue in the same way for other special cases, and the special cases depend on the function. E.g., for `sin` you could define a special treatment if t[0] is 0, \pi/3, pi/2, etc, but those exceptions would change for `cos`.

The solution is simply what we said before, `sin` should be implemented for `Double` (and `Single`) within `DoubleDouble`. Perhaps using the crmlib for the floating part would be ok.

**Comment by dpsanders** [comment]
I don't agree, I think `0` is a very special case. 
My idea was to use the `TaylorSeries` package to *calculate* `sin` for `DoubleDouble`, e.g.
```
julia> using TaylorSeries, DoubleDouble

julia> t = Taylor(10)
ERROR: UndefVarError: Taylor not defined

julia> t = Taylor1(10)
 1.0 t + 𝒪(t¹¹)

julia> evaluate(sin(t), Double(0.5))
Double(0.4794255386164159, -6.5227459551767006e-18)
 - value: 0.47942553861641589621459841260216
```
But this currently doesn't work, since the coefficients of `sin(t)` are `Float64`; I need them to be `Double`.

**Comment by dpsanders** [comment]
Proof of concept at #59

**Comment by dpsanders** [comment]
It works!
```
julia> t = Taylor1(Double{Float64}, 100)
 Double(1.0, 0.0)
 - value: 1 t + 𝒪(t¹⁰¹)

julia> evaluate(exp(t), Double(0.5))
Double(1.6487212707001282, -4.731568479435833e-17)
 - value: 1.6487212707001281468486507878142

julia> exp(big"0.5")
1.648721270700128146848650787814163571653776100710148011575079311640661021194214
```

**Comment by lbenet** [comment]
Indeed, zero is a special case. 

But this should be defined in the implementation of `sin` for `Double`, and not in general in `sin` for `Taylor1`. The reason that everything works smoothly for `Taylor1{Interval{T}}` is because there is a method `sin(a::Interval)` defined in ValidatedNumerics!

What i have in mind is:
```julia
function sin{T}(x::Double{T})
    zero(x) == x && return zero(x)
     ... # rest of the implementation
end
```

This should allow the recurrence that it is used in `Taylor1` to work, at least when you are expanding around zero.

**Comment by dpsanders** [comment]
I don't agree, for the reason that I wrote above -- I want to use `TaylorSeries` to calculate the coefficients of the expansion of `sin` for a new number type, *before* I have defined `sin` for that number type.
This seems like a very reasonable thing to do.

**Comment by lbenet** [comment]
I understand the circular argument here. For me, the question is, if I create `MySuperbType{T} <: Number`, does `TaylorSeries` should adapt to its possible existence, or rather should it be the module where `MySuperbType` is introduced adapted? I think the second should happen.

**Comment by dpsanders** [comment]
`TaylorSeries` should not have to be *changed* for a new number type; it should just be generic enough to work with any new number type. And this turns out to be easy to do, so we should do it IMO.

**Comment by lbenet** [comment]
> `TaylorSeries` should not have to be changed for a new number type; it should just be generic enough to work with any new number type. And this turns out to be easy to do, so we should do it IMO.

I agree with that statement. Yet, I think that *adapting*  TaylorSeries is precisely not following that statement. (Though, indeed, it is not a change so difficult in the present case).

**Comment by lbenet** [comment]
In my opinion, the problem arises because there are some methods lacking for `Double`. The following is a simple example, which is important for `sin(a::Double)`:

```julia
julia> dd = Double(0.0)
Double(0.0, 0.0)
 - value: 0

julia> 1 * dd
ERROR: MethodError: `convert` has no method matching convert(::Type{DoubleDouble.Double{Float64}}, ::Int64)
This may have arisen from a call to the constructor DoubleDouble.Double{Float64}(...),
since type constructors fall back to convert methods.
Closest candidates are:
  call{T}(::Type{T}, ::Any)
  convert{T<:Real}(::Type{T<:Real}, ::Complex{T<:Real})
  convert{T<:Number}(::Type{T<:Number}, ::Char)
  ...
 in * at promotion.jl:168
```

The point is that, if I have this method working, I can build up a *proof of concept* which lies completely on DoubleDouble, and uses TaylorSeries as intended.

**Comment by lbenet** [comment]
Multiplying a `Int` by a `Double` is required in line 758 of `Taylor1.jl`; if this works, i can *easily* create `sin` for a `Double` by a Taylor expansion around zero, just the way you propose.

**Comment by lbenet** [comment]
The following code extends some methods of `DoubleDouble.jl` making that `sin` and `cos` work without touching TaylorSeries. The price to pay is to require `TaylorSeries.jl` in `DoubleDouble.jl`.

```julia
using TaylorSeries, DoubleDouble
import Base: convert, promote_rule
import Base: one, zero, sin, cos

convert{T<:AbstractFloat}(::Type{Double{T}}, k::Integer) = Double(convert(T,k))
promote_rule{T<:Number}(::Type{Double{T}}, Real) = Double{promote_type(T, S)}

zero{T<:AbstractFloat}(::Double{T}) = Double(zero(T))
zero{T<:AbstractFloat}(::Type{Double{T}}) = Double(zero(T))
one{T<:AbstractFloat}(::Double{T}) = Double(one(T))
one{T<:AbstractFloat}(::Type{Double{T}}) = Double(one(T))

function sin{T<:AbstractFloat}(a::Double{T})
    a == zero(T) && return a
    # Use TaylorSeries
    t = taylor1_variable(Double{T}, 20)
    sin_t = sin(t)
    return evaluate(sin_t, a)
end

function cos{T<:AbstractFloat}(a::Double{T})
    a == zero(T) && return one(a)
    # Use TaylorSeries
    t = taylor1_variable(Double{T}, 20)
    cos_t = cos(t)
    return evaluate(cos_t, a)
end
```

Then:
```julia
julia> sin(Double(0.5))
Double(0.479425538604203, -5.103969869884509e-18)
 - value: 0.47942553860420300027328792588708

julia> sin(BigFloat(0.5))
4.794255386042030002732879352155713880818033679406006751886166131255350002878161e-01

julia> cos(Double(0.5))
Double(0.8775825618903728, -4.262314986406798e-17)
 - value: 0.87758256189037271611628158281585

julia> cos(BigFloat(0.5))
8.775825618903727161162815826038296519916451971097440529976108683159507632742139e-01
```

So it also works!

**Comment by lbenet** [comment]
Better results are probably obtained for even higher-order Taylor expansions... Using 100 for the Taylor expansions yields

```julia
julia> sin(Double(0.5))
Double(0.479425538604203, -5.103969860556012e-18)
 - value: 0.47942553860420300027328793521557

julia> sin(BigFloat(0.5))
4.794255386042030002732879352155713880818033679406006751886166131255350002878161e-01

julia> cos(Double(0.5))
Double(0.8775825618903728, -4.2623149864279997e-17)
 - value: 0.87758256189037271611628158260383

julia> cos(BigFloat(0.5))
8.775825618903727161162815826038296519916451971097440529976108683159507632742139e-01
```
which is even better.

**Comment by lbenet** [comment]
Moreover, an implementation like this allows to carefully choose some values around which the expansion is done, to reduce the order of the Taylor series used. A trivial example is `pi` or `pi/2`.

**Comment by dpsanders** [comment]
Yes, this is exactly what I had in mind.
Note that you don't need to explicitly define `zero` and `one` once you have a suitable `convert` method, since there are generic fallbacks!

**Comment by lbenet** [comment]
I guess you are right regarding that it is only needed some plausible convert methods. 

The point is that all this should be defined for `Double`; the Taylor expansion then works automatically!

**Comment by dpsanders** [comment]
This also allows the code to *almost* work for #51, which would be great.

**Comment by lbenet** [comment]
I guess you are right, as long as you are expanding around zero, such that the return value of the evaluation of the zero-th order coefficient is, e.g., rational. 

Aside from #51, if you define some "table" of return values for specific zero-order evaluations, the series can be shortened a lot, and precision can be better controlled.

**Comment by dpsanders** [comment]
I am now worried about type instability. Maybe these functions should return a Taylor object with elements of the same type as the elements of its input and not try to be clever about finding out the

**Comment by lbenet** [comment]
So far, the functions are type stable. In general, for doing an expansion around some value `t0`, I think `f(t0)` should define the type of the returned `Taylor1`-series, which is what is done currently. You can always convert to whatever else:
```julia
julia> sin(Taylor1(10))
 1.0 t - 0.16666666666666666 t³ + 0.008333333333333333 t⁵ - 0.0001984126984126984 t⁷ + 2.7557319223985893e-6 t⁹ + 𝒪(t¹¹)

julia> convert(Taylor1{Rational{Int}},ans)
 1//1 t - 1//6 t³ + 1//120 t⁵ - 1//5040 t⁷ + 1//362880 t⁹ + 𝒪(t¹¹)
```

Once said this, I think that you could specify the initial conversion when defining the function for the specific type, say `sin` for `Double`. For instance, by defining that `sin(Double(0.0))` returns `0//1`.

**Comment by dpsanders** [comment]
I don't agree. For example, consider the following where the `convert` is just wrong:
```
julia> t = Taylor1(Rational{Int128}, 30)
 1//1 t + 𝒪(t³¹)

julia> exp(t)
 1//1 + 1//1 t + 1//2 t² + 1//6 t³ + 1//24 t⁴ + 1//120 t⁵ + 1//720 t⁶ + 1//5040 t⁷ + 1//40320 t⁸ + 1//362880 t⁹ + 1//3628800 t¹⁰ + 1//39916800 t¹¹ + 1//479001600 t¹² + 1//6227020800 t¹³ + 1//87178291200 t¹⁴ + 1//1307674368000 t¹⁵ + 1//20922789888000 t¹⁶ + 1//355687428096000 t¹⁷ + 1//6402373705728000 t¹⁸ + 1//121645100408832000 t¹⁹ + 1//2432902008176640000 t²⁰ + 1//51090942171709440000 t²¹ + 1//1124000727777607680000 t²² + 1//25852016738884976640000 t²³ + 1//620448401733239439360000 t²⁴ + 1//15511210043330985984000000 t²⁵ + 1//403291461126605635584000000 t²⁶ + 1//10888869450418352160768000000 t²⁷ + 1//304888344611713860501504000000 t²⁸ + 1//8841761993739701954543616000000 t²⁹ + 1//265252859812191058636308480000000 t³⁰ + 𝒪(t³¹)

julia> t = Taylor1(Float64, 30)
 1.0 t + 𝒪(t³¹)

julia> convert(Taylor1{Rational{Int128}}, exp(t))
 1//1 + 1//1 t + 1//2 t² + 1//6 t³ + 1//24 t⁴ + 1//120 t⁵ + 1//720 t⁶ + 1//5040 t⁷ + 1//40320 t⁸ + 1//362880 t⁹ + 1//3628800 t¹⁰ + 1//39916800 t¹¹ + 1//479001550 t¹² + 1//6227012191 t¹³ + 1//87176603682 t¹⁴ + 1//1307294779224 t¹⁵ + 1//20826036441708 t¹⁶ + 1//329652014863266 t¹⁷ + 1//1 t¹⁸ + 1//1 t¹⁹ + 1//1 t²⁰ + 1//1 t²¹ + 1//1 t²² + 1//1 t²³ + 1//1 t²⁴ + 1//1 t²⁵ + 1//1 t²⁶ + 1//1 t²⁷ + 1//1 t²⁸ + 1//1 t²⁹ + 1//1 t³⁰ + 𝒪(t³¹)
```

**Comment by dpsanders** [comment]
Basically I don't agree that the package should choose the type of the result for me -- it would be much more transparent to just use the type that I provided in the original `Taylor1` object, and throw an error if I tried something like `sin(3 + Taylor1(Rational{Int}, 10))`

**Comment by lbenet** [comment]
I agree that there are inconsistencies. They are related to the fact that round errors occur for `Float64`s (and a priori not for `Rational`s), and `rationalize` somehow exposes them. 

You are thinking on Taylor series *around zero* where you can write the rational value nicely for many functions (but a priori not all!), but the package allows to develop around *any* value, and the function evaluated at the point where you are expanding may not be a rational. (I think this feature makes the package pretty general.) If you are interested in `sin(pi*1.1/4)` it makes sense to develop around `pi/4`, despite `sin(pi/4)=1/sqrt(2)` is not a rational.

While not perfect, the `convert` methods precisely allow the user to choose the type `T` of a `Taylor1{T}` object.

Also, I think this is the way Julia actually reacts: `sin(0//1)` returns `0.0` and not `0//1`.

**Comment by lbenet** [comment]
Also, things are also subtle among rationals:

```julia
julia> rationalize(Int128,sqrt(2))
77227930//54608393

julia> rationalize(Int128,1/sqrt(2))
54608393//77227930

julia> convert(Rational{Int128},sqrt(2))
6369051672525773//4503599627370496

julia> convert(Rational{Int128},1/sqrt(2))
1592262918131443//2251799813685248
```

**Comment by dpsanders** [comment]
Yes, converting *from* a float *to* a rational is rarely a good idea; converting *from* a rational *to* a float is much better behaved.

**Comment by dpsanders** [comment]
The point is that the machinery for calculating the coefficients with an arbitrary type, in particular the exact coefficients as rationals, already exists in the package for expansions around 0. I feel that this should be exposed somehow, *without* the need to define `sin` etc. for the other type.

**Comment by dpsanders** [comment]
I suppose one possibility would be a syntax like `sin(Rational{Int}, t)`

**Comment by lbenet** [comment]
We are consistent with the (default) behavior in Julia: `sqrt(4//1)` yields `2.0` and not `2//1`; similarly, `sin(0//1)` yields `0.0` and not `0//1`.

With regards to your [last comment](https://github.com/JuliaDiff/TaylorSeries.jl/issues/56#issuecomment-229490726), I think that that should be implemented in Base. I agree that that may be useful; I simply don't know what to expect for `sin(Rational{Int}, pi/4)`, as shown above.

**Comment by dpsanders** [comment]
I see your point, but we should definitely provide the possibility to access the *exact* coefficients (rationals) as well as inexact ones (floats). Maybe a better syntax is something like
`sin(t::Taylor1; output_type=Rational{Int})` with a keyword argument.

**Comment by lbenet** [comment]
I think we already do, with the caveats that are included with the `convert` methods and concretely `rationalize`. I think you are too focused on expansions around 0 which maybe nice, but what about expanding `sin` around `pi/4`?

**Comment by dpsanders** [comment]
I have just pushed a proof of concept to #59 . This actually leads to *simpler* code, and allows the following, which I think is pretty nice

```
julia> using TaylorSeries

julia> t = Taylor1(Rational{Int}, 10)
 1//1 t + 𝒪(t¹¹)

julia> exp(t)
 1.0 + 1.0 t + 0.5 t² + 0.16666666666666666 t³ + 0.041666666666666664 t⁴ + 0.008333333333333333 t⁵ + 0.001388888888888889 t⁶ + 0.0001984126984126984 t⁷ + 2.48015873015873e-5 t⁸ + 2.7557319223985893e-6 t⁹ + 2.7557319223985894e-7 t¹⁰ + 𝒪(t¹¹)

julia> exp(t, Rational{Int})
 1//1 + 1//1 t + 1//2 t² + 1//6 t³ + 1//24 t⁴ + 1//120 t⁵ + 1//720 t⁶ + 1//5040 t⁷ + 1//40320 t⁸ + 1//362880 t⁹ + 1//3628800 t¹⁰ + 𝒪(t¹¹)
```

**Comment by dpsanders** [comment]
We cannot do anything about `sin` around `pi/4`, since we can't even represent `pi/4` exactly...

**Comment by dpsanders** [comment]
Simon Byrne just pointed out that `sinpi` is an option for this.

**Comment by lbenet** [comment]
I agree that it looks nicer. What happens for an expansion around `pi/4`? 

> We cannot do anything about sin around pi/4, since we can't even represent pi/4 exactly...

*All* irrational numbers, and a large amount of rationals have issues with the `Float64` representation. Yet, the question is fair and an answer is expected. Returning a different type depending on the value is type unstable.

Using `Rational{Int}` and a long enough expansion of `sin(t)`, what do you get evaluating it at `pi/4`?

**Comment by lbenet** [comment]
Just for the record, gp/pari which uses rational arithmetic as far as it can, returns for `sin(0)` a float.

**Comment by dpsanders** [comment]
I haven't tried to implement `sin` yet. For `exp`, expanding around some other point gives rational representations of the floating-point coefficients. This doesn't seem particularly useful to me -- the useful case is for expansion around 0:

```
julia> t = Taylor1(Rational{Int128}, 10)
 1//1 t + 𝒪(t¹¹)

julia> exp(3+t, Rational{Int128})
 2826788018837635//140737488355328 + 2826788018837635//140737488355328 t + 2826788018837635//281474976710656 t² + 2826788018837635//844424930131968 t³ + 2826788018837635//3377699720527872 t⁴ + 565357603767527//3377699720527872 t⁵ + 565357603767527//20266198323167232 t⁶ + 565357603767527//141863388262170624 t⁷ + 565357603767527//1134907106097364992 t⁸ + 565357603767527//10214163954876284928 t⁹ + 565357603767527//102141639548762849280 t¹⁰ + 𝒪(t¹¹)
```

**Comment by lbenet** [comment]
Then, why don't you define that `sin(Double(0.0))` returns `0//1`, or `exp(Double(0.0))` returns `1//1`; see [this comment](https://github.com/JuliaDiff/TaylorSeries.jl/issues/56#issuecomment-229168671). Then you get what you want automatically.

**Comment by dpsanders** [comment]
No, because I want to have rationals sometimes, and `Double`s at other times. This has nothing to do with `Double`s; it is also true with `Float64`. Of course it makes no sense to define `sin(Double(0.0))` to return a rational, as you pointed out.

**Comment by dpsanders** [comment]
In other words, I need a way to *control* the output type.

**Comment by lbenet** [comment]
`convert` :-)

**Comment by dpsanders** [comment]
I've already shown why `convert` isn't good enough...

**Comment by dpsanders** [comment]
I actually like the new solution I proposed in #59

**Comment by lbenet** [comment]
The point is that in #59 you have modified the expansion of `exp` around *any* point, to one around zero (`coeffs[1] = one(output_type)`). This in my opinion reduces generality, instead of increasing it, 

Moreover, in applications such as Taylor integration, this imposes a very large order of the expansion if you want to get some convergence of the series. Put differently, using an expansion around zero, how large the order must be to get a good approximation of `exp(3.1)`? It's simply better to expand around 3!

**Comment by lbenet** [comment]
If, at all, I think you should define *new* methods, which involve explicitly (not optionally) the type of output, e.g., `exp{T}(a::Taylor1{T}, output_type)`. This does extend the way you want `exp`, without altering the usual use of `exp` for a `Taylor1` polynomial.

**Comment by lbenet** [comment]
Sorry, I meant `exp{T}(a::Taylor1{T}, output_type::Type)`, or something alike.

**Comment by dpsanders** [comment]
Oh, looks like I didn't push the latest version, sorry.


On Wed, Jun 29, 2016 at 7:08 PM, Luis Benet <notifications@github.com>
wrote:

> Sorry, I meant exp{T}(a::Taylor1{T}, output_type::Type), or something
> alike.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaDiff/TaylorSeries.jl/issues/56#issuecomment-229516245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALtTgulOpGbYxADWt_zQSvBNwWiP5fiks5qQvrjgaJpZM4I_WMQ>
> .
>



-- 
Dr. David P. Sanders

Profesor Titular "B" / Associate Professor
Departamento de Física, Facultad de Ciencias
Universidad Nacional Autónoma de México (UNAM)

dpsanders@g <dpsanders@ciencias.unam.mx>mail.com / Twitter: @DavidPSanders
<https://twitter.com/DavidPSanders>
http://sistemas.fciencias.unam.mx/~dsanders / GitHub: dpsanders
<https://github.com/dpsanders>

Cubículo / office: #414, 4o. piso del Depto. de Física

Tel.: (+52 55) 5622 4965

**Comment by lbenet** [comment]
I think the new version is type unstable: for `t=Taylor1(10)`, `exp(t, Rational{Int})` and `exp(1.0+t, Rational{Int})` will return `Taylor1{Rational{Int}}` and `Taylor1{Float64}`. Am I right?

My proposal is to have two *distinct* methods: `exp(a::Taylor1)` and `exp{T}(a::Taylor1{T}, output_type::Type)`; the second one has no default `output_type` to avoid collisions with the first one, and emphasizing that the user should know what is doing (expanding around a given point).

**Comment by lbenet** [comment]
Probably narrowing the `output_type` to something reasonable would be better...

**Comment by dpsanders** [comment]
No, they both return the same type:
```
julia> t = Taylor1(3)
 1.0 t + 𝒪(t⁴)

julia> exp(t, Rational{Int})
 1//1 + 1//1 t + 1//2 t² + 1//6 t³ + 𝒪(t⁴)

julia> exp(1.0+t, Rational{Int})
 6121026514868073//2251799813685248 + 6121026514868073//2251799813685248 t + 6121026514868073//4503599627370496 t² + 2040342171622691//4503599627370496 t³ + 𝒪(t⁴)
```
but with order 10 the rationals overflow for the second one.

**Comment by dpsanders** [comment]
I'm not sure what you mean by "narrowing `output_type` to something reasonable.

**Comment by lbenet** [comment]
I guess you are right regarding the supposed type unstability, or perhaps is a type change. Simply, `one(output_type)` can be a different type from `exp(zeroth_order)`, precisely depending onthe value of `zeroth_order` (lines 623 and 625).

The proposal of having two methods, in which the second allows for `output_type` argument, is ensure that `output_type::Type{T<:Number}`.

**Comment by dpsanders** [comment]
`output_type` is defined (in the optional argument) as `float(T)`, which is always the type returned by `exp(zeroth_order)`, since `exp(x::Real)` is defined as `exp(float(x))` (c.f. `@edit exp(3)`).

So everything is consistent and there is no type instability. I like this approach, since it makes the output type explicit, instead of relying implicitly on inferring the type as `exp(zeroth_order)` (even though the result is the same).

**Comment by lbenet** [comment]
After a long silence, I've spent some time playing with this PR, since it has an attractive side, despite my reluctance.

(A minor comment: for this PR to work you need to be in "convert_reals" branch of DoubleDouble.jl, which has some `convert` methods that make `float(Double{Float64})` and `zero(Double{Float64})` work.)

The whole idea here is to avoid the evaluation of `exp(a.coeffs[1])` ---which yields an stack overflow error---. Instead, it uses the fact that `exp(Double(0.0))` must be `Double(1.0)` and then expand around zero, with the hope that enough terms suffice to approximate. As you showed, the idea works. So, e.g., it can be used to define `exp(::Double{T})`.

However, I still have some doubts which keep me reluctant. At the moment, if you use this approach to calculate `exp(Double(53.5))` you need about 180 Taylor coefficients to have an error on the last displayed digit. The way to know that you need this many is by comparing against the `BigFloat` result, or by comparing results by increasing the number of coefficients. In either case, this calculation is longer (~5x) than the calculation with bigfloats.

Another point is, for example, if you need the Taylor expansion around some point t0. That is, `exp(t0+t)`, with `t` a `Taylor1{T}` variable, and `T` the type of `t0`. For this you need `exp(t0)`, so we are at the beginning.

Summarizing my point of view: 
I like the idea of `output_type`, though I think it should be a keyword parameter. I don't like the idea of expanding around zero as a special case; maybe it is better
